plugins/searchonline.koplugin/main.lua

```diff
@@ -0,0 +1,313 @@
+    disabled_default_engines = G_reader_settings:readSetting("searchonline_disabled_default_engines", {}),
+}
+
+function SearchOnline:init()
+    if self.ui and self.ui.menu then
+        self.ui.menu:registerToMainMenu(self)
+    end
+    self.ui.highlight:addToHighlightDialog("12_onlinesearch", function(reader_highlight_instance)
+        return self:createHighlightMenuItem(reader_highlight_instance)
+    end)
+
+    -- Combine defaults with the user search engines
+    self:rebuildSearchEngines()
+end
+
+function SearchOnline:rebuildSearchEngines()
+    -- Build the local search_engines list from defaults
+    self.search_engines = {}
+    for _, engine in ipairs(self.default_search_engines) do
+        if not self.disabled_default_engines[engine] then
+            table.insert(self.search_engines, engine)
+        end
+    end
+    -- Add user search engines
+    for _, engine in ipairs(self.user_search_engines) do
+        table.insert(self.search_engines, engine)
+    end
+end
+
+
+function SearchOnline:getHighlight(reader_highlight_instance)
+    if not reader_highlight_instance.selected_text or not reader_highlight_instance.selected_text.text then
+        return
+    end
+
+    local query = util.cleanupSelectedText(reader_highlight_instance.selected_text.text)
+    return query
+end
+
+function SearchOnline:createHighlightMenuItem(reader_highlight_instance)
+    return {
+        text = _("Search online"),
+        enabled = Device:canOpenLink(),
+        callback = function()
+            NetworkMgr:runWhenOnline(function()
+                self:chooseSearch(self:getHighlight(reader_highlight_instance))
+            end)
+        end
+    }
+end
+
+-- combines the query string with the URL string of the chosen
+-- search engine and passes it to `Device:openLink()`
+function SearchOnline:searchOnline(query, search_engine)
+    
+    local chosen_engine = nil
+    for _, engine in ipairs(self.search_engines) do
+        if engine == search_engine then
+            chosen_engine = engine
+            break
+        end
+    end
+    if not chosen_engine then
+        HandleUrlString:urlError(search_engine .. " not found")
+        return
+    end
+
+    if not chosen_engine:find("{q}") then
+        HandleUrlString:urlError("The URL is misformatted, put {q} where the search term should go: " .. chosen_engine)
+        return
+    end
+
+    local sanitised_query = url.escape(query)
+    local build_url = chosen_engine:gsub("{q}", sanitised_query)
+
+    if HandleUrlString:isValidUrl(build_url) then
+        Device:openLink(build_url)
+    end
+end
+
+
+function SearchOnline:chooseSearch(query)
+    local radio_buttons = {}
+    for index, engine_url in ipairs(self.search_engines) do
+        table.insert(radio_buttons, { -- build a button for each
+            {                         -- saved search engine
+                text = HandleUrlString:getDomainName(engine_url),

```

Comment:
Feels a bit odd to have to call this getDisplayName extraction each time we display the button - usually, we'd have this stored as self.something and rebuild by a dedicated function when it is updated.